src/Functions/printf.cpp

```diff
@@ -0,0 +1,343 @@
+#include <Columns/ColumnFixedString.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnStringHelpers.h>
+#include <DataTypes/DataTypeString.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Functions/formatString.h>
+#include <IO/WriteHelpers.h>
+
+#include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <fmt/format.h>
+#include <fmt/printf.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int ILLEGAL_COLUMN;
+extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+namespace
+{
+
+class FunctionPrintf : public IFunction
+{
+private:
+    ContextPtr context;
+    FunctionOverloadResolverPtr function_concat;
+
+    struct Instruction
+    {
+        std::string_view format;
+        size_t rows;
+        bool is_literal; /// format is literal string without any argument
+        ColumnWithTypeAndName input; /// Only used when is_literal is false
+
+        ColumnWithTypeAndName execute() const
+        {
+            if (is_literal)
+                return executeLiteral(format);
+            else if (isColumnConst(*input.column))
+                return executeConstant(input);
+            else
+                return executeNonconstant(input);
+        }
+
+        [[maybe_unused]] String toString() const
+        {
+            std::ostringstream oss; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+            oss << "format:" << format << ", rows:" << rows << ", is_literal:" << is_literal << ", input:" << input.dumpStructure()
+                << std::endl;
+            return oss.str();
+        }
+
+    private:
+        ColumnWithTypeAndName executeLiteral(std::string_view literal) const
+        {
+            ColumnWithTypeAndName res;
+            auto str_col = ColumnString::create();
+            str_col->insert(fmt::sprintf(literal));
```

Comment:
`sprintf` also supports positional arguments. While not supporting such cases is fine (error message can be better), but for some cases result may be not expected (when only '%1$' is used)

```
SELECT printf('%1$d %2$d', 10 * number, number)
FROM numbers(11)

Elapsed: 0.002 sec. 

Received exception from server (version 24.7.1):
Code: 1001. DB::Exception: Received from localhost:9000. DB::Exception: fmt::v9::format_error: argument not found. (STD_EXCEPTION)


---
SELECT printf('%1$d %1$d', 10 * number, number)
FROM numbers(2)

   ┌─printf('%1$d %1$d', multiply(10, number), number)─┐
1. │ 0 0                                               │
2. │ 10 1                                              │
   └───────────────────────────────────────────────────┘
```

Reply:
Actually error message can be improved for all errors:

```
select printf('%d %.f',  number, number) from numbers(2);

Received exception from server (version 24.7.1):
Code: 1001. DB::Exception: Received from localhost:9000. DB::Exception: fmt::v9::format_error: invalid type specifier. (STD_EXCEPTION)
```

Maybe we can `try ... catch` at least?

Reply:
Also some format specifiers looks a it dangerous, for example
```
n
    The argument shall be a pointer to an integer into which is written the number of bytes written to the output so far by this call to one of the fprintf() functions. No argument is converted.
```

https://pubs.opengroup.org/onlinepubs/009695399/functions/fprintf.html

Even though we don't pass pointers there, and `fmt` will throw exception instead of using number as a pointer, still can be a concert and maybe worth validating and allowing only limited set of modifiers?

Reply:
> Actually error message can be improved for all errors:
> 
> ```
> select printf('%d %.f',  number, number) from numbers(2);
> 
> Received exception from server (version 24.7.1):
> Code: 1001. DB::Exception: Received from localhost:9000. DB::Exception: fmt::v9::format_error: invalid type specifier. (STD_EXCEPTION)
> ```
> 
> Maybe we can `try ... catch` at least?

done. 

``` sql 
SELECT printf('%d %.f', number, number)
FROM numbers(2)

Query id: 47c3543b-ee85-4d56-85b5-2e768e550ef2


Elapsed: 0.024 sec. 

Received exception from server (version 24.7.1):
Code: 36. DB::Exception: Received from localhost:9001. DB::Exception: Bad format %.f in function printf with __table1.number UInt64 UInt64(size = 2) as input argument, detail: invalid type specifier: while executing 'FUNCTION printf('%d %.f'_String :: 1, __table1.number : 2, __table1.number :: 2) -> printf('%d %.f'_String, __table1.number, __table1.number) String : 0'. Stack **trace:**
```

Reply:
> `sprintf` also supports positional arguments. While not supporting such cases is fine (error message can be better), but for some cases result may be not expected (when only '%1$' is used)
> 
> ```
> SELECT printf('%1$d %2$d', 10 * number, number)
> FROM numbers(11)
> 
> Elapsed: 0.002 sec. 
> 
> Received exception from server (version 24.7.1):
> Code: 1001. DB::Exception: Received from localhost:9000. DB::Exception: fmt::v9::format_error: argument not found. (STD_EXCEPTION)
> 
> 
> ---
> SELECT printf('%1$d %1$d', 10 * number, number)
> FROM numbers(2)
> 
>    ┌─printf('%1$d %1$d', multiply(10, number), number)─┐
> 1. │ 0 0                                               │
> 2. │ 10 1                                              │
>    └───────────────────────────────────────────────────┘
> ```

It is pity that current framework doesn't support positional arguments. But the error message is improved already.

Reply:
> Also some format specifiers looks a it dangerous, for example
> 
> ```
> n
>     The argument shall be a pointer to an integer into which is written the number of bytes written to the output so far by this call to one of the fprintf() functions. No argument is converted.
> ```
> 
> https://pubs.opengroup.org/onlinepubs/009695399/functions/fprintf.html
> 
> Even though we don't pass pointers there, and `fmt` will throw exception instead of using number as a pointer, still can be a concert and maybe worth validating and allowing only limited set of modifiers?

I tested it and we should not worry about it for what you said above. In CH function `printf`, only string/fixedstring/float/int columns are allowed to be arguments. Any unmatched type specifier and corresponding column will cause an exception.  
```
SELECT printf('%n', number)
FROM numbers(10)

Query id: 17b1aad1-9313-4778-a646-f627621c62ef


Elapsed: 0.003 sec. 

Received exception from server (version 24.7.1):
Code: 36. DB::Exception: Received from localhost:9001. DB::Exception: Bad format %n in function printf with __table1.number UInt64 UInt64(size = 10) as input argument, detail: invalid type specifier: while executing 'FUNCTION printf('%n'_String :: 1, __table1.number :: 2) -> printf('%n'_String, __table1.number) String : 0'. Stack trace:

```